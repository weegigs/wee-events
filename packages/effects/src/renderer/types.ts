import * as Effect from "effect/Effect";
import { Entity, Payload } from "@weegigs/events-core";

export interface RenderFailure {
  readonly description: string;
}

export type Renderer<R, P extends Payload, A extends Payload> = Effect.Effect<(entity: Entity<P>) => Effect.Effect<A, RenderFailure>, never, R>;
export namespace Renderer {
  export type IO<P extends Payload, A extends Payload = Payload> = Renderer<unknown, P, A>;
}

export type AnyRenderer<R> = Renderer<R, Payload, Payload>;
export namespace AnyRenderer {
  export type IO = AnyRenderer<unknown>;

  export const from = <R, P extends Payload, A extends Payload>(source: Renderer<R, P, A>): AnyRenderer<R> => {
    return source as unknown as AnyRenderer<R>;
  };
}
