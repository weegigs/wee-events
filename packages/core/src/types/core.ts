import * as z from "zod";

export type Primitive = string | number | boolean;
export type Value = Primitive | Payload | List;
export type EmptyPayload = { [member: string]: never };
export type Payload = { [member: string]: Value } | EmptyPayload;
export type List = Primitive[] | Payload[] | List[];

export namespace Payload {
  const primitive = z.union([z.string(), z.number(), z.boolean()]);
  const list: z.Schema<List> = z.lazy(() => z.union([z.array(primitive), z.array(payload), z.array(list)]));
  const value: z.Schema<Value> = z.lazy(() => z.union([primitive, list, payload]));
  const payload: z.Schema<Payload> = z.lazy(() => z.record(value));

  export const schema = payload;
  export const parse = (value: unknown): Payload => schema.parse(value);
}

export namespace EmptyPayload {
  export const schema = z.record(z.never());
}

export type AggregateId<Type extends string = string> = {
  type: Type;
  key: string;
};

export namespace AggregateId {
  export const schema = <Type extends string>(type: Type): z.Schema<AggregateId<Type>> =>
    z.object({
      type: z.literal(type),
      key: z.string().min(1),
    }) as unknown as z.Schema<AggregateId<Type>>;

  export const encode = (aggregate: AggregateId): string => `${aggregate.type}.${aggregate.key}`;
  export const decoder =
    (type: string) =>
    (aggregate: string): AggregateId => {
      const [_type, ...key] = aggregate.split(".");
      return schema(type).parse({ type: _type, key: key.join(".") });
    };

  export const parser =
    <Type extends string>(type: Type) =>
    (value: unknown) =>
      schema(type).parse(value);

  export function create<const T extends string>(type: T, key: string): AggregateId<T> {
    return { type, key } as const;
  }
}

export interface DomainEvent<Type extends string = string, Data extends Payload = Payload> {
  type: Type;
  data: Data;
}

export namespace DomainEvent {
  export const schema = <S extends z.Schema>(type: string, data: S) =>
    z.object({
      type: z.literal(type),
      data,
    });
}

export interface RecordedEvent<E extends DomainEvent = DomainEvent> {
  /**
   * The aggregate that events belongs to.
   */
  aggregate: AggregateId;

  /**
   * The revision for the aggregate.
   */
  revision: string;

  /**
   * Unique identifier representing this event. ULID format.
   */
  id: string;

  /**
   * Type of this event.
   */
  type: E["type"];

  /**
   * Representing when this event was created in the database system.
   */
  timestamp: string;

  /**
   * Data of this event.
   */
  data: E["data"];

  /**
   * Representing the metadata associated with this event.
   */
  metadata: RecordedEvent.Metadata;
}

export namespace RecordedEvent {
  export type Metadata = {
    causationId?: string;
    correlationId?: string;
  };
}

export type Revision = string;

export namespace Revision {
  export type Initial = "00000000000000000000000000";
  export const Initial = "00000000000000000000000000";

  export const isInitial = (revision: Revision): boolean => revision === Initial;

  export const schema = z.string().min(26).max(26);
}
