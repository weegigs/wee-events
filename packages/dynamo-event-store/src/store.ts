import _ from "lodash";
import * as z from "zod";

import { DynamoDBClient } from "@aws-sdk/client-dynamodb";
import { DynamoDBDocumentClient, QueryCommand, TransactWriteCommand, QueryCommandOutput } from "@aws-sdk/lib-dynamodb";

import { ChangeSet } from "./changeset";

// Simple error classification functions to replace deprecated @aws-sdk/service-error-classification
const isRetryableByTrait = (error: any) => error?.retryable === true;
const isClockSkewError = (error: any) => error?.name === "ClockSkew";
const isThrottlingError = (error: any) => error?.name === "ThrottlingException" || error?.name === "ProvisionedThroughputExceededException";
const isTransientError = (error: any) => error?.name === "ServiceUnavailable" || error?.name === "InternalServerError";

import { retry } from "@weegigs/events-common";
import { Cypher, Tokenizer, no } from "@weegigs/events-cypher";

import {
  DomainEvent,
  EventStore,
  RecordedEvent,
  Revision,
  AggregateId,
  ExpectedRevisionConflictError,
  RevisionConflictError,
  Payload,
} from "@weegigs/events-core";

export namespace DynamoEventStore {
  export interface Options {
    client: () => DynamoDBDocumentClient;
    cypher: Cypher;
    tokenizer: Tokenizer;
  }
}

const defaults: DynamoEventStore.Options = {
  client: () =>
    DynamoDBDocumentClient.from(new DynamoDBClient({}), {
      marshallOptions: {
        removeUndefinedValues: true,
      },
    }),
  cypher: no.cypher,
  tokenizer: no.tokenizer,
};

type LatestRecord = {
  pk: string;
  sk: string;
  revision: string;
  timestamp: string;
};

function latestFor(changeSet: ChangeSet): LatestRecord {
  return {
    pk: changeSet.pk,
    sk: `latest-revision`,
    revision: changeSet.revision,
    timestamp: changeSet.timestamp,
  };
}

function latestCondition(changeset: ChangeSet, expectedVersion?: string) {
  if (undefined === expectedVersion) {
    return {
      ConditionExpression: "attribute_not_exists(#revision) Or #revision < :revision",
      ExpressionAttributeNames: { "#revision": "revision" },
      ExpressionAttributeValues: { ":revision": changeset.revision },
    };
  }

  if (expectedVersion === Revision.Initial) {
    return {
      ConditionExpression: "attribute_not_exists(#revision)",
      ExpressionAttributeNames: { "#revision": "revision" },
    };
  }

  return {
    ConditionExpression: "#revision = :revision",
    ExpressionAttributeNames: { "#revision": "revision" },
    ExpressionAttributeValues: { ":revision": expectedVersion },
  };
}

const isRetryableError = (error: unknown) => {
  return isRetryableByTrait(error as any) || isClockSkewError(error as any) || isThrottlingError(error as any) || isTransientError(error as any);
};

const isConditionCheckFailed = (error: unknown) =>
  (error as any)?.name === "TransactionCanceledException" &&
  (error as any).CancellationReasons?.[0]?.Code === "ConditionalCheckFailed";

export class DynamoEventStore implements EventStore {
  private readonly $client: DynamoDBDocumentClient;
  private readonly $table: string;
  private readonly $encoder: ChangeSet.EventEncoder;
  private readonly $decoder: ChangeSet.ChangeSetDecoder;

  constructor(table: string, options: Partial<DynamoEventStore.Options> = {}) {
    const { client, cypher, tokenizer } = { ...defaults, ...options };

    this.$client = client();
    this.$table = table;
    this.$encoder = ChangeSet.encoder(cypher, tokenizer);
    this.$decoder = ChangeSet.decoder(cypher);
  }

  private $read = async (
    key: string,
    decrypt: boolean,
    _after?: Revision,
    cursor: any = undefined
  ): Promise<{ events: RecordedEvent[]; next: any }> => {
    const query = new QueryCommand({
      TableName: this.$table,
      ScanIndexForward: true,
      ExclusiveStartKey: cursor,
      KeyConditionExpression: "#pk = :aggregate And begins_with(#sk, :change_set_prefix)",
      ExpressionAttributeNames: {
        "#pk": "pk",
        "#sk": "sk",
      },
      ExpressionAttributeValues: {
        ":aggregate": key,
        ":change_set_prefix": "change-set#",
      },
    });

    const { Items, LastEvaluatedKey } = await retry(
      async () => this.$client.send(query),
      (e: Error | QueryCommandOutput) => {
        return _.isError(e) ? isRetryableError(e) : false;
      },
      { limit: 5, delay: 37 }
    );

    const events: RecordedEvent[][] = await Promise.all(
      (Items ?? [])
        .map((item: any) => ChangeSet.schema.parse(item))
        .map((changeSet) => this.$decoder(changeSet, decrypt))
    );

    return { events: events.flat(), next: LastEvaluatedKey };
  };

  load = async (aggregate: AggregateId, options: EventStore.LoadOptions = {}): Promise<RecordedEvent[]> => {
    const key = AggregateId.encode(aggregate);
    const { decrypt, afterRevision: after } = { decrypt: false, ...options };

    let events: RecordedEvent[] = [];
    let cursor: any | undefined = undefined;
    do {
      const { events: page, next } = await retry(
        () => this.$read(key, decrypt, after, cursor),
        (result) => _.isError(result) && isRetryableError(result)
      );

      events = events.concat(page);
      cursor = next;
    } while (cursor);

    return events;
  };

  publish = async (
    aggregate: AggregateId,
    events: DomainEvent | DomainEvent[],
    options: EventStore.PublishOptions = {}
  ): Promise<Revision> => {
    const _events = Array.isArray(events) ? events : [events];
    if (_events.length === 0) {
      throw new Error(`expected at least one event to be published`);
    }

    const { expectedRevision } = options;

    try {
      return await retry(
        async (): Promise<Revision> => this.$publish(aggregate, _events, options),
        (e: Error | Revision) => {
          if (!_.isError(e)) {
            return false;
          }
          if (
            e.name === "TransactionCanceledException" &&
            (e as any).CancellationReasons?.[0]?.Code === "ConditionalCheckFailed"
          ) {
            return expectedRevision === undefined;
          }

          return isRetryableError(e);
        },
        { limit: 5, delay: 37 }
      );
    } catch (e: unknown) {
      if (_.isError(e) && isConditionCheckFailed(e)) {
        if (expectedRevision !== undefined) {
          throw new ExpectedRevisionConflictError(expectedRevision);
        }

        throw new RevisionConflictError();
      }

      throw e;
    }
  };

  private async $publish(
    aggregate: AggregateId,
    events: DomainEvent[],
    options: EventStore.PublishOptions = {}
  ): Promise<Revision> {
    const { encrypt, expectedRevision } = options;

    const recorded: RecordedEvent[] = await Promise.all(
      events
        .map((event) => {
          // Validate that the event has the proper DomainEvent structure
          const domainEventSchema = z.object({
            type: z.string().min(1),
            data: Payload.schema
          });
          return domainEventSchema.parse(event);
        })
        .map((event) =>
          this.$encoder(aggregate, event, _.pick(options, "correlationId", "causationId"), encrypt ?? false)
        )
    );

    const changeSet = await ChangeSet.create(aggregate, recorded);
    return this.$write(changeSet, expectedRevision);
  }

  private async $write(changeSet: ChangeSet, expectedRevision?: string): Promise<Revision> {
    const TableName = this.$table;

    const latest = latestFor(changeSet);
    const latestCheck = latestCondition(changeSet, expectedRevision);

    const write = new TransactWriteCommand({
      TransactItems: [
        {
          Put: {
            TableName,
            Item: latest,
            ...latestCheck,
          },
        },
        {
          Put: {
            TableName,
            Item: changeSet,
          },
        },
      ],
    });

    await this.$client.send(write);

    return changeSet.revision;
  }
}
