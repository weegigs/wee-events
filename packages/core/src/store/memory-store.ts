import { DateTime } from "luxon";
import { monotonicFactory } from "ulid";

import {
  AggregateId,
  DomainEvent,
  RecordedEvent,
  Revision,
} from "../types";
import { EventStore } from "./types";

const ulid = monotonicFactory();

export class MemoryStore implements EventStore {
  aggregates: Record<string, RecordedEvent[]> = {};

  clear() {
    this.aggregates = {};
  }

  async load(aggregate: AggregateId): Promise<RecordedEvent[]> {
    return this.aggregates[AggregateId.encode(aggregate)] ?? [];
  }

  async publish(aggregate: AggregateId, events: DomainEvent | DomainEvent[]): Promise<Revision> {
    const current = this.aggregates[AggregateId.encode(aggregate)] ?? [];
    const revision = current[current.length - 1]?.revision ?? Revision.Initial;

    const e = Array.isArray(events) ? events : [events];
    if (e.length == 0) {
      return revision;
    }

    const recorded: RecordedEvent[] = e.map((event) => {
      const revision = ulid()
      const timestamp = DateTime.now().toISO()

      if (timestamp == null) {
        throw new Error("unexpected invalid timestamp")
      }

      return {
        ...event,
        id: revision,
        revision,
        aggregate: aggregate,
        timestamp,
        metadata: {},
      };
    });

    const updated = current.concat(...recorded);
    this.aggregates[AggregateId.encode(aggregate)] = updated;

    return updated[updated.length - 1].revision;
  }
}
