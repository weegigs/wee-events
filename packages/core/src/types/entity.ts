import * as z from "zod";

import { AggregateId, Payload, Revision } from "./core";

export interface Entity<T extends Payload = Payload> {
  aggregate: AggregateId;
  type: string;
  revision: Revision;
  state: T;
}

export namespace Entity {
  export const schema = <Type extends string, Data extends z.ZodTypeAny>(type: Type, state: Data) =>
    z.object({
      aggregate: AggregateId.schema(type),
      type: z.string().min(1),
      revision: Revision.schema,
      state,
    });

  export function create<State extends Payload>(
    aggregate: AggregateId,
    type: string,
    revision: Revision,
    state: State
  ): Entity<State> {
    return { aggregate, type, revision, state } as const;
  }
}
