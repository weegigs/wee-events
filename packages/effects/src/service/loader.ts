import * as T from "@effect-ts/core/Effect";
import { Has } from "@effect-ts/core/Has";

import * as store from "../event-store";

import * as wee from "@weegigs/events-core";

import { z } from "zod";
import { pipe } from "@effect-ts/core";
import { DomainEvent, Payload } from "@weegigs/events-core";

type State = wee.Payload;

export type Reducer<A extends Payload, Type extends string, Data extends Payload> = (
  state: A,
  event: wee.RecordedEvent<DomainEvent<Type, Data>>
) => A;

export type Initializer<A extends Payload, Type extends string, Data extends Payload> = (
  event: wee.RecordedEvent<DomainEvent<Type, Data>>
) => A;

export type Init<A extends State> = (id: wee.AggregateId) => A;

export class EntityNotAvailableError extends Error {
  constructor(public readonly aggregate: wee.AggregateId, public readonly type: string) {
    super(`Entity of type ${type} not available from aggregate ${wee.AggregateId.encode(aggregate)}`);
    this.name = "EntityNotAvailableError";

    Object.setPrototypeOf(this, EntityNotAvailableError.prototype);
  }
}

interface InitializedLoader<A extends State> {
  reducer: <Type extends string, Data extends Payload>(
    event: Type,
    schema: z.Schema<Data>,
    reducer: Reducer<A, Type, Data>
  ) => InitializedLoader<A>;

  make(): EntityLoader<A>;
}

interface InitializingLoader<A extends State> extends InitializedLoader<A> {
  initializer<Type extends string, Data extends Payload>(
    event: Type,
    schema: z.Schema<Data>,
    initializer: Initializer<A, Type, Data>
  ): InitializingLoader<A>;
}

export interface EntityLoader<A extends State> {
  load(
    id: wee.AggregateId
  ): T.Effect<Has<store.EventLoader>, store.LoaderError | EntityNotAvailableError, wee.Entity<A>>;
  events(): Record<string, z.Schema>;
}

export interface EntityDescription<A extends State> {
  readonly name: string;
  readonly events: Record<string, z.Schema>;
  readonly schema: z.Schema<A>;
}

type Initializers<A extends State> = Record<
  string,
  { schema: z.Schema<unknown>; initializer: Initializer<A, string, Payload> }
>;

type Reducers<A extends State> = Record<string, { schema: z.Schema<unknown>; reducer: Reducer<A, string, Payload> }>;

export namespace EntityLoader {
  const $make = <A extends State>(loader: $Loader<A>): EntityLoader<A> => {
    const { init, initializers, reducers, type } = loader;

    const replay = (events: wee.RecordedEvent[], state?: A): [A | undefined, wee.Revision] => {
      let result = state;
      let revision = wee.Revision.Initial;

      for (const event of events) {
        if (result === undefined) {
          const { initializer } = initializers[event.type] ?? { initializer: undefined };
          if (initializer !== undefined) {
            result = initializer(event);
            revision = event.revision;
          }
        } else {
          const { reducer } = reducers[event.type] ?? { reducer: undefined };
          if (reducer !== undefined) {
            result = reducer(result, event);
            revision = event.revision;
          }
        }
      }

      return [result, revision];
    };

    const events: () => Record<string, z.Schema> = () => {
      return {};
    };

    const load = (id: wee.AggregateId) =>
      pipe(
        T.do,
        T.bind("events", () => store.load(id)),
        T.bind("entity", ({ events }) =>
          T.suspend(() => {
            const [state, revision] = replay(events, typeof init == "function" ? init(id) : undefined);
            if (state === undefined) {
              return T.fail(new EntityNotAvailableError(id, type));
            }
            const entity: wee.Entity<A> = {
              aggregate: id,
              type,
              state,
              revision,
            };

            return T.succeed(entity);
          })
        ),
        T.map(({ entity }) => entity)
      );

    return { load, events };
  };

  class $Loader<A extends State> implements InitializedLoader<A> {
    constructor(
      readonly type: string,
      readonly init: (aggregate: wee.AggregateId) => A | undefined,
      readonly initializers: Initializers<A>,
      readonly reducers: Reducers<A>
    ) {}

    reducer<Type extends string, Data extends Payload>(
      event: Type,
      schema: z.Schema<Data>,
      reducer: Reducer<A, Type, Data>
    ): InitializedLoader<A> {
      return new $Loader<A>(this.type, this.init, this.initializers, {
        ...this.reducers,
        [event]: { schema, reducer: reducer as Reducer<A, string, Payload> },
      });
    }

    make(): EntityLoader<A> {
      return $make(this);
    }
  }

  class $Initializing<A extends State> extends $Loader<A> implements InitializingLoader<A> {
    constructor(type: string, initializers: Initializers<A>) {
      super(type, () => undefined, initializers, {});
    }
    initializer<Type extends string, Data extends wee.Payload>(
      event: Type,
      schema: z.ZodType<Data, z.ZodTypeDef, Data>,
      initializer: Initializer<A, Type, Data>
    ): InitializingLoader<A> {
      return new $Initializing<A>(this.type, {
        ...this.initializers,
        [event]: { schema, initializer: initializer as Initializer<A, string, Payload> },
      });
    }
  }

  export const init = <A extends State>(type: string, init: Init<A>): InitializedLoader<A> => {
    const loader = new $Loader<A>(type, init, {}, {});
    return loader;
  };

  export const initializer = <A extends State, Type extends string, Data extends Payload>(
    type: string,
    event: Type,
    schema: z.ZodType<Data, z.ZodTypeDef, Data>,
    initializer: Initializer<A, Type, Data>
  ): InitializingLoader<A> => {
    return new $Initializing<A>(type, {
      [event]: { schema, initializer: initializer as Initializer<A, string, Payload> },
    });
  };
}
