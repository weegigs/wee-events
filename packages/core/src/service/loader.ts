import _ from "lodash";
import { z } from "zod";

import * as store from "../store";
import { AggregateId, DomainEvent, Entity, Payload, RecordedEvent, Revision } from "../types";
import { State } from "./state";

export interface EntityLoader<A extends State> {
  load(id: AggregateId): Promise<Entity<A>>;
}

export type EntityDescription<S extends State> = { type: string; schema: z.ZodType<S, z.ZodTypeDef, S> };

export type LoaderDescription<S extends State> = {
  entity(): EntityDescription<S>;
  events(): Record<string, z.Schema>;

  create: (store: store.EventStore) => EntityLoader<S>;
};

export type Reducer<A extends State, Event extends DomainEvent> = (state: A, event: RecordedEvent<Event>) => A;
export type Initializer<A extends State, Event extends DomainEvent> = (event: RecordedEvent<Event>) => A;

export class EntityNotAvailableError extends Error {
  constructor(public readonly aggregate: AggregateId, public readonly type: string) {
    super(`Entity of type ${type} not available from aggregate ${AggregateId.encode(aggregate)}`);
    this.name = "EntityNotAvailableError";

    Object.setPrototypeOf(this, EntityNotAvailableError.prototype);
  }
}

export namespace LoaderDescription {
  type Initializers<A extends State> = Record<
    string,
    { schema: z.Schema<unknown>; initializer: Initializer<A, DomainEvent> }
  >;

  type Reducers<A extends State> = Record<string, { schema: z.Schema<unknown>; reducer: Reducer<A, DomainEvent> }>;

  class Components<A extends State> {
    constructor(
      readonly entity: EntityDescription<A>,
      readonly init: ((aggregate: AggregateId) => A) | undefined,
      readonly initializers: Initializers<A>,
      readonly reducers: Reducers<A>
    ) {}
  }

  const $create = <A extends State>(components: Components<A>): LoaderDescription<A>["create"] => {
    return (store: store.EventStore) => {
      const {
        init,
        initializers,
        reducers,
        entity: { type },
      } = components;

      const replay = (events: RecordedEvent[], state?: A): [A | undefined, Revision] => {
        let result = state;
        let revision = Revision.Initial;

        for (const event of events) {
          if (result === undefined) {
            const { initializer } = initializers[event.type] ?? { initializer: undefined };
            if (initializer !== undefined) {
              result = initializer(event);
              if (result !== undefined) {
                revision = event.revision;
              }
            }
          } else {
            const { reducer } = reducers[event.type] ?? { reducer: undefined };
            if (reducer !== undefined) {
              result = reducer(result, event);
              revision = event.revision;
            }
          }
        }

        return [result, revision];
      };

      const load = async (id: AggregateId) => {
        const events = await store.load(id);
        const [state, revision] = replay(events, typeof init == "function" ? init(id) : undefined);
        if (state === undefined) {
          throw new EntityNotAvailableError(id, type);
        }
        const entity: Entity<A> = {
          aggregate: id,
          type,
          state,
          revision,
        };

        return entity;
      };

      return { load };
    };
  };

  function $events<A extends State>(components: Components<A>): Record<string, z.Schema> {
    const { initializers, reducers } = components;

    return _.merge(
      {},
      ...Object.entries(initializers).map(([event, { schema }]) => ({ [event]: schema })),
      ...Object.entries(reducers).map(([event, { schema }]) => ({ [event]: schema }))
    );
  }

  function $make<A extends State>(components: Components<A>): LoaderDescription<A> {
    const { entity } = components;

    const create = $create(components);
    const events = $events(components);

    return { entity: () => entity, events: () => events, create };
  }

  interface Initialized<A extends State> {
    description(): LoaderDescription<A>;
    reducer<Type extends string, Data extends Payload>(
      event: Type,
      schema: z.Schema<Data>,
      reducer: Reducer<A, DomainEvent<Type, Data>>
    ): Initialized<A>;
  }

  class $Initialized<A extends State> extends Components<A> implements Initialized<A> {
    reducer<Type extends string, Data extends Payload>(
      event: Type,
      schema: z.Schema<Data>,
      reducer: Reducer<A, DomainEvent<Type, Data>>
    ): Initialized<A> {
      return new $Initialized<A>(this.entity, this.init, this.initializers, {
        ...this.reducers,
        [event]: { schema, reducer: reducer as Reducer<A, DomainEvent> },
      });
    }

    description(): LoaderDescription<A> {
      return $make(this);
    }
  }

  interface Initializing<A extends State> extends Initialized<A> {
    initializer<Type extends string, Data extends Payload>(
      event: Type,
      schema: z.ZodType<Data, z.ZodTypeDef, Data>,
      initializer: Initializer<A, DomainEvent<Type, Data>>
    ): Initializing<A>;
  }

  class $Initializing<A extends State> extends $Initialized<A> implements Initializing<A> {
    initializer<Type extends string, Data extends Payload>(
      event: Type,
      schema: z.ZodType<Data, z.ZodTypeDef, Data>,
      initializer: Initializer<A, DomainEvent<Type, Data>>
    ): $Initializing<A> {
      return new $Initializing<A>(
        this.entity,
        this.init,
        {
          ...this.initializers,
          [event]: { schema, initializer: initializer as Initializer<A, DomainEvent> },
        },
        this.reducers
      );
    }
  }

  export const fromInitFunction = <A extends State>(
    entity: { type: string; schema: z.ZodType<A, z.ZodTypeDef, A> },
    init: (aggregate: AggregateId) => A
  ): Initialized<A> => {
    return new $Initialized(entity, init, {}, {});
  };

  export const fromInitializer = <A extends State, Type extends string, Data extends Payload>(
    entity: { type: string; schema: z.ZodType<A, z.ZodTypeDef, A> },
    event: Type,
    schema: z.ZodType<Data, z.ZodTypeDef, Data>,
    initializer: Initializer<A, DomainEvent<Type, Data>>
  ): Initializing<A> => {
    return new $Initializing<A>(
      entity,
      undefined,
      {
        [event]: { schema, initializer: initializer as Initializer<A, DomainEvent> },
      },
      {}
    );
  };
}
