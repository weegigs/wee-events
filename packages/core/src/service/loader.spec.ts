import { DateTime } from "luxon";
import { decodeTime, monotonicFactory } from "ulid";
import { z } from "zod";

import { EventStore } from "../store";
import { DomainEvent, RecordedEvent, Revision } from "../types";
import { EntityNotAvailableError, Initializer, LoaderDescription, Reducer } from "./loader";

const ulid = monotonicFactory();

namespace Calculator {
  export const schema = z.object({
    total: z.number(),
  });
}
type Calculator = z.infer<typeof Calculator.schema>;

type Added = DomainEvent<"added", { amount: number }>;
namespace Added {
  export const schema = z.object({ amount: z.number() });
}

const addr: Reducer<Calculator, Added> = (state, event) => {
  return { ...state, total: state.total + event.data.amount };
};

const addi: Initializer<Calculator, Added> = (event) => {
  return { total: event.data.amount };
};

type Subtracted = DomainEvent<"subtracted", { amount: number }>;
namespace Subtracted {
  export const schema = z.object({ amount: z.number() });

  export const reducer: Reducer<Calculator, Subtracted> = (state, event) => ({
    ...state,
    total: state.total - event.data.amount,
  });

  export const initializer: Initializer<Calculator, Subtracted> = (event) => ({ total: -event.data.amount });
}

describe("loaders", () => {
  const load = jest.fn();

  const store: EventStore = {
    publish: jest.fn(),
    load,
  };

  const tid = { type: "test", key: "1" };
  const added = (amount: number): RecordedEvent<Added> => {
    const id = ulid();
    const timestamp = DateTime.fromMillis(decodeTime(id)).toISO()!;

    return {
      id,
      aggregate: tid,
      revision: id,
      type: "added",
      data: { amount },
      timestamp,
      metadata: {},
    };
  };
  const subtracted = (amount: number): RecordedEvent<Subtracted> => {
    const id = ulid();
    const timestamp = DateTime.fromMillis(decodeTime(id)).toISO()!;

    return {
      id,
      aggregate: tid,
      revision: id,
      type: "subtracted",
      data: { amount },
      timestamp,
      metadata: {},
    };
  };

  describe("with init function", () => {
    const description = LoaderDescription.fromInitFunction<Calculator>(
      { type: "calculator", schema: Calculator.schema },
      () => ({ total: 0 })
    )
      .reducer("added", Added.schema, addr)
      .description();

    const loader = description.create(store);

    it("should return an initialized entity if no events are available", async () => {
      load.mockResolvedValueOnce([]);

      const result = await loader.load(tid);

      expect(result.state).toEqual({ total: 0 });
      expect(result).toMatchSnapshot();
    });

    it("should return a upto date entity if events are available", async () => {
      load.mockResolvedValueOnce([added(1), added(2), added(3)]);

      const result = await loader.load(tid);

      expect(result.state).toEqual({ total: 6 });
      expect(result.revision).not.toEqual(Revision.Initial);
    });

    it("should provide a description of the events handled", () => {
      const events = description.events();

      expect(Object.keys(events)).toHaveLength(1);
      expect(events["added"]).toBeDefined();
    });
  });

  describe("with initializer", () => {
    const description = LoaderDescription.fromInitializer(
      { type: "calculator", schema: Calculator.schema },
      "added",
      Added.schema,
      addi
    )
      .initializer("subtracted", Subtracted.schema, Subtracted.initializer)
      .reducer("added", Added.schema, addr)
      .reducer("subtracted", Subtracted.schema, Subtracted.reducer)
      .description();

    const loader = description.create(store);

    it("should throw an EntityNotAvailable error if no events are available", async () => {
      load.mockResolvedValueOnce([]);

      await expect(loader.load(tid)).rejects.toThrowError(EntityNotAvailableError);
    });

    it("should return a upto date entity if events are available", async () => {
      load.mockResolvedValueOnce([added(1), subtracted(2), added(3)]);

      const result = await loader.load(tid);

      expect(result.state).toEqual({ total: 2 });
      expect(result.revision).not.toEqual(Revision.Initial);
    });

    it("should handle alternative initializers", async () => {
      load.mockResolvedValueOnce([subtracted(3), added(2), added(3)]);

      const result = await loader.load(tid);

      expect(result.state).toEqual({ total: 2 });
      expect(result.revision).not.toEqual(Revision.Initial);
    });

    it("should provide a description of the events handled", () => {
      const events = description.events();

      expect(Object.keys(events)).toHaveLength(2);
      expect(events["added"]).toBeDefined();
      expect(events["subtracted"]).toBeDefined();
    });
  });
});
