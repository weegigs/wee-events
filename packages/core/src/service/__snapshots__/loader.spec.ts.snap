// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loaders with init function should return an initialized entity if no events are available 1`] = `
{
  "aggregate": {
    "key": "1",
    "type": "test",
  },
  "revision": "00000000000000000000000000",
  "state": {
    "total": 0,
  },
  "type": "calculator",
}
`;
